### PR TITLE
Add codegen for pure-functions

### PR DESCRIPTION
**Pull request by junjihashimoto**
* Codegen of higer level functions of dynamic tensor
* Not generate overloaded functions.

### CODE REVIEW

**Comment by austinvhuang** [comment]
Saves a lot of work, and I suppose we can have a complementary module for a few exceptions / custom interfaces.

It looks like it looks like there's some business logic for explicit exclusions + picking one among alternative instances. What are the rationale / heuristics? Looks like there's some sort order prioritization between instances.

**Review by apaszke — COMMENTED** [review]
While this is a good idea, I think that we should only codegen ops specified in a whitelist that should also let us remap their names. Something like:
```hs
data Binding = UnsafeBinding { cName :: String, hsName :: String }

bindRename = UnsafeBinding
bind name = UnsafeBinding name name

bindings = [
  bind "relu",
  bindRename "sum" "sumAll"
]
```

The only difficulty might be the need to specify overloads (like with `sum`), but I think that would make for a much nicer and cleaner interface.

**Comment by junjihashimoto** [comment]
@austinvhuang @apaszke 
Thank you for reviewing.

> What are the rationale / heuristics? 

I am thinking those.
There are some ideas.
1, Remove the functions writing to the arguments, e.g. functions having suffix of "_out" or "_" like asin_out or add_.
    https://github.com/pytorch/pytorch/tree/3b1c3996e1c82ca8f43af9efa196b33e36efee37/aten/src/ATen/native
2, Remove the functions having prefix of "_"
3, Remove the functions having not supported types(e.g. Generator) on higher level now
    (It is wrong that ScalarType is removed on current implementation.) 


I will make the remapper.

**Comment by austinvhuang** [comment]
How about this proposal - code is generated by default (using the above heuristics or some variant of them), reserving the `bindings` specification for remappings and explicit exclusions (exclusion could use a new `bindRemove` function or alternatively `hsName` could be `Maybe String`).

In other words, the thousands of lines of "bind" entries are the default and don't have to be explicitly specified, but if a "bindRename" or "bindRemove" exists for the function that overrides the implicit "bind" entry.

Rationale:
- Instead of maintaining a `bindings` variable spec that is comparable in length to the generated code, only need to specify with the small fraction of special case functions requiring a name remapping or manually written binding. 
- Changes upstream will be propagated by default for most cases without having to determine the upstream delta in new functions and manually adding/deleting entries in the remapper

Downsides:
- Once in a while if there's an upstream change that breaks the assumptions of this process or heuristics, that may result in broken output and/or require rework of the code generation

In spite of that one downside, it seems like the maintenance burden would be less with this route. What do you think?

**Comment by junjihashimoto** [comment]
I agree with you.
To specify overloads (like with sum), it is useful binding target can be specified by both c++'s name and haskell's one.
So I think following the format of bindings and the spec-file(yaml). 

--On haskell--
```
data BindingName
 = HsName String
  |  CppName String

data Binding
 = BindRename  { src :: BindingName, hsName :: String }
 | Bind { src :: BindingName}
 | BindRemove{ src :: BindingName}

instance FromJSON BindingName
instance FromJSON Binding

bindRename = BindRename  
bind = Bind 

bindings = [
  bind (CppName "relu"),
  bindRename (HsName "sum_t") "sumAll"
]
```

--On yaml--
```
- bind:        {src: {cpp_name: relu}} # bind is default, so this line is not needed.
- bind_rename: {src: {hs_name: sum_t}, hs_name: sumAll}
- bind_remove: {src: {cpp_name: add}}
```

**Comment by austinvhuang** [comment]
Given binding rename is ideally a limited set of special cases and not a large chunk of data, my vote would be to have that as just Haskell so there's not another layer of indirection.

**Comment by junjihashimoto** [comment]
> Given binding rename is ideally a limited set of special cases and not a large chunk of data, my vote would be to have that as just Haskell so there's not another layer of indirection.

**Comment by junjihashimoto** [comment]
I just want to save our work,
I do not want to disturb writing in haskell.

**Comment by junjihashimoto** [comment]
Above binding format is changed. It supports only haskell's name.

File: codegen/src/RenderPure.hs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+      let l = nativeFunctionsFilter fns bnd
+
+      case notUniqList (map fst l) of
+        [] -> do 
+          T.writeFile (basedir <> "/Torch/Functions/Native.hs") $
+            template "Torch.Functions.Native" $
+            renderFunctions l
+        xs -> do
+          putStrLn "---Duplicated functions are as follows. ----"
+          forM_ xs $ \x -> do
```
Comment (junjihashimoto) [inline codegen/src/RenderPure.hs:135]: When codegen finds duplicated definitions of functions like sum_t and sum_tt, it stops generating codes and outputs the names of functions.

File: hasktorch/src/Torch/Tensor.hs [hunk 2]
```diff
@@ -221,3 +225,23 @@ instance Show Tensor where
...
+    ptr_list <- mapM (\x -> (cast x return :: IO (ForeignPtr ATen.Tensor))) xs
+    cast ptr_list f
+  uncast xs f = uncast xs $ \ptr_list -> do
+    tensor_list <- mapM (\(x :: ForeignPtr ATen.Tensor) -> uncast x return) ptr_list
+    f tensor_list
+
+
+instance Castable String (ForeignPtr ATen.StdString) where
+  cast str f = ATen.newStdString_s str >>= f
+  uncast xs f = ATen.string_c_str xs >>= f
```
Comment (junjihashimoto) [inline hasktorch/src/Torch/Tensor.hs]: `instance Castable String (ForeignPtr ATen.StdString)` should be in ATen.Cast.
I will move it.

File: hasktorch/src/Torch/Functions/Native.hs [hunk 3]
```diff
@@ -0,0 +1,1157 @@
...
+hardshrink_backward _grad_out _self _lambd = unsafePerformIO $ (cast3 ATen.hardshrink_backward_tts) _grad_out _self _lambd
+
+rsqrt :: Tensor -> Tensor
+rsqrt _self = unsafePerformIO $ (cast1 ATen.rsqrt_t) _self
+
+celu :: Tensor -> Float -> Tensor
+celu _self _alpha = unsafePerformIO $ (cast2 ATen.celu_ts) _self _alpha
+
+detach :: Tensor -> Tensor
+detach _self = unsafePerformIO $ (cast1 ATen.detach_t) _self
```
Comment (junjihashimoto) [inline hasktorch/src/Torch/Functions/Native.hs]: `detach` always needs IO-monad. So it should be removed.

**Comment by austinvhuang** [comment]
@junjihashimoto this is going to be a hugely useful when it lands, do you think this is rdy / close? 

Side note - if the codegen here replicates (or provides a close equivalent) functionality to the hand-coded operations in `Functions.hs` I'm fine with deprecating the `Functions.hs` version in favor of these instead of adding those to exclusions just because we happened to get around to those first.

It's okay if there's some rework needed of examples or tests as a result if it simplifies the API and makes it more coherent.

**Comment by junjihashimoto** [comment]
@austinvhuang Thank you for your comments.
I think this is partially okay. 
I am thinking which function-name is better snake-case or caml-case? and how to generate factory-methods like randint.
But we can add the other things after this PR. 
For now I merge it.

**Pull request merged**